### PR TITLE
fix(inference): sample via Gumbel-max exponential race in fp32

### PR DESCRIPTION
**Pull request by sidsingh-nvidia**
## What

Replaces `torch.multinomial` in `TorchSampling` with an exponential-race (Gumbel-max) draw, and computes three quantities in fp32 that were previously done in the logits dtype.

| | before | after |
|---|---|---|
| sampling | `torch.multinomial(probs)` | `argmax(probs / q)`, `q ~ Exp(1)` |
| top-p softmax/cumsum | bf16 | fp32 |
| sampling softmax | bf16 | fp32 |
| reported `log_softmax` | logits dtype | fp32 |

## Why

On a 131k-entry vocabulary, `softmax().cumsum()` in bf16 saturates before the tail is accumulated, so `cumulative_probs > top_p` becomes true for a block of low-probability entries. `torch.multinomial` then walks an inverse CDF, so the same cumulative-sum rounding erodes the tail a second time at sampling time.

The exponential race is distributionally identical to multinomial but never forms a cumulative sum. It is also exactly what vLLM does ([`topk_topp_sampler.py::random_sample`](https://github.com/vllm-project/vllm/blob/main/vllm/v1/sample/ops/topk_topp_sampler.py)), so the two engines' sampling paths agree — relevant for Megatron-inference-vs-vLLM parity work. vLLM additionally uses it to avoid the CPU-GPU sync `torch.multinomial` incurs.

## Measurements

24 degeneracy-prone prompts x 48 samples = 1,152 generations per arm. nanov3.5, `temperature=1.0`, `top_p=1.0`, `max_tokens=24,576`, identical prompts posted as raw token ids.

| arm | mean | median | p90 | p95 |
|---|---|---|---|---|
| as-trained (multinomial, bf16) | 5202 | 2780 | 13689 | 23966 |
| + dtype fixes only | 5182 | 2831 | 13712 | 22402 |
| **+ exponential race (this PR)** | **4878** | 2755 | **11490** | **19785** |
| vLLM 0.25.1 | 4824 | 2711 | 11379 | 19107 |

Median is unchanged, so this affects only the tail. The dtype fixes alone move nothing (mean −20 tokens, p90 +23); the algorithm change closes ~95% of the p90 gap to vLLM and leaves the arm statistically indistinguishable from it (mean-length difference +54 tokens, p = 0.73).

Sampler correctness was checked directly against `torch.multinomial` at N = 400k: total variation to the true distribution 0.0134 vs 0.0141, and in the far tail (entries below 1e-4·p_max, true mass 0.003472) the race recovers 0.003595 vs multinomial's 0.003385.

## Caveat

Paired by prompt, the generation-length improvement is **p ≈ 0.08** — suggestive, not significant. The distributional argument and the N=400k tail check are the stronger evidence; the length numbers are supporting context.

## Note for reviewers

This is ported from an internal branch. That branch also tightened the top-p gate from `elif top_p > 0.0` to `elif 0.0 < top_p < 1.0`, because at `top_p = 1.0` the bf16 cumsum masked ~7,680 of 131,072 entries — nucleus filtering ran when none was requested. **That hunk is deliberately omitted here**, since `main` already covers it more cleanly via `is_no_op_top_p()`.

The branch is cut from a slightly older `main`; the changed lines do not overlap the `is_no_op_top_p` refactor, so it merges cleanly. Happy to rebase.

### CODE REVIEW

File: megatron/core/inference/sampling/torch_sampling.py [hunk 1]
```diff
@@ -111,8 +111,18 @@ def sample_from_logits(
...
         )
-        probabilities = filtered.softmax(dim=-1)
-        sampled = torch.multinomial(probabilities, num_samples=1, generator=generator).view(-1)
+        probabilities = filtered.softmax(dim=-1, dtype=torch.float32)
+        # Gumbel-max in exponential form: draw q ~ Exp(1) per vocabulary entry and
+        # take argmax(p / q). Distributionally identical to `torch.multinomial`,
+        # but it never forms a cumulative sum, so the low-probability tail is not
+        # eroded by the rounding an inverse-CDF walk accumulates across a 100k+
+        # vocabulary. Matches vLLM's sampler
+        # (`vllm/v1/sample/ops/topk_topp_sampler.py::random_sample`), which keeps
```
Comment (santhnm2) [inline megatron/core/inference/sampling/torch_sampling.py:120]: Can we drop a permalink here so we have a static reference?

**Review by santhnm2 — APPROVED** [review]

**Review by kvareddy — APPROVED** [review]

File: megatron/core/inference/sampling/torch_sampling.py (hunk 1 above)
Comment (sidsingh-nvidia) [reply megatron/core/inference/sampling/torch_sampling.py]: https://github.com/vllm-project/vllm/blob/7702ee87dba0d8eed7f201e77a0a6613aac738a8/vllm/v1/sample/ops/topk_topp_sampler.py#L464

Comment (sidsingh-nvidia) [reply megatron/core/inference/sampling/torch_sampling.py]: Added the permalink to the comment in 1f5910f86: https://github.com/vllm-project/vllm/blob/7702ee87dba0d8eed7f201e77a0a6613aac738a8/vllm/v1/sample/ops/topk_topp_sampler.py#L464

**Comment by sidsingh-nvidia** [comment]
/ok to test [1f5910f](https://github.com/NVIDIA/Megatron-LM/pull/7182/commits/1f5910f86b8494549b659dea88eb73d3176e1fe1)

**Review by ericharper — APPROVED** [review]

**Comment by sidsingh-nvidia** [comment]
Investigated the 3 failing \`tests/unit_tests\` CI jobs from the run on 1f5910f86 (before the branch was later updated with \`main\`, merge commit a8a7d9282):

**1. \`tests/unit_tests/inference/engines/test_dynamic_engine.py\` — golden-value staleness, fixed (c04ab9d4c).**
All 8 failing parametrizations (\`test_simple[*-hybrid]\`, every \`num_cuda_graphs\` x \`InferenceCudaGraphScope\` combo) hit the exact same, single mismatch: \`mamba_expected_generated_tokens[4]\`, position 10, expected \`35\` got \`73\`. The Gumbel-max sampler this PR introduces is distributionally identical to \`torch.multinomial\` but consumes the RNG stream differently, so a fixed seed now produces a different (still valid, in-range) draw at that one position. No other position/request/parametrization differs, and the \`-gpt\` parametrizations (which don't exercise this golden array) all passed — consistent with an RNG-draw change, not a regression. Updated the golden value.

**2. \`tests/unit_tests/test_optimizer.py::test_distributed_optimizer_reload_main_params_from_fused_mla_canonical_state_dict\` — pre-existing on \`main\`, not touched.**
Distributed-optimizer / MFSDP canonical state-dict reload test; no plausible path from a sampling-kernel change to this failure. Not fixed here.

**3. \`tests/unit_tests/models/test_hybrid_moe_model.py::TestHybridMoEModel::test_constructor\` — \`Failed: Mamba MoE config drift detected!\`, pre-existing on \`main\`, not touched.**
Config-drift snapshot check for Mamba MoE model construction; also unrelated to sampling. Notably, \`main\`'s history includes a recent \`ci: Revert recently merged changes as they cause CI unit test failures (#7326)\` immediately followed by \`Add implementation for ShortcutMoE (#6959)\` touching this same area — this looks like a fresh, currently-unresolved break on \`main\` itself, not something introduced by this branch.

I could not re-run the full distributed/multi-GPU CI suite locally (these jobs use \`torch.distributed.elastic\` across multiple H100/GB200 workers, not available in this environment) -- items 2 and 3 are classified by commit-history/subject-matter reasoning, not by reproducing green on \`main\` locally. Worth a maintainer double-checking #2 and #3 are indeed already-known main breakage before merging.

**Comment by sidsingh-nvidia** [comment]
/ok to test [c04ab9d](https://github.com/NVIDIA/Megatron-LM/pull/7182/commits/c04ab9d4c74e6893a651118e0b6be86f5b7d1804)

**Review by sylvesterkaczmarek — COMMENTED** [review]
`log_probs_kernel()` still returns the logits dtype. `log_probs` is allocated with `torch.empty_like(logits)`, so the new fp32 `log_softmax` is immediately cast back to bf16/fp16 on assignment. That defeats the precision change for returned log-probs. Please allocate `log_probs` as float32 (or otherwise preserve the fp32 result) and add a dtype regression for low-precision logits.

**Review by sylvesterkaczmarek — CHANGES_REQUESTED** [review]
Rechecked current `c04ab9d4`. The returned log-probability precision change is still ineffective: `log_probs_kernel()` allocates `log_probs` with `torch.empty_like(logits)`, then assigns the fp32 `torch.log_softmax(logits.float(), ...)` result into that low-precision buffer. For bf16/fp16 logits this immediately casts the values back down. Please allocate the returned log-probs as fp32 (or otherwise preserve the fp32 result) and add a dtype/value regression for low-precision logits.

**Comment by sidsingh-nvidia** [comment]
/ok to test [cb0d507](https://github.com/NVIDIA/Megatron-LM/pull/7182/commits/cb0d507fe804a814133d83a3b37c150b1d8812ab)

**Comment by sidsingh-nvidia** [comment]
Thanks @sylvesterkaczmarek, addressed.

**Comment by sidsingh-nvidia** [comment]
/ok to test [ddd4c6d](https://github.com/NVIDIA/Megatron-LM/pull/7182/commits/ddd4c6dba5bda8a561d46c5d9873943d2cd986d1)

**Review by sylvesterkaczmarek — APPROVED** [review]
Rechecked current `ddd4c6db`. The returned buffer is now explicitly float32, and the new unit coverage exercises bf16, fp16, and fp32 logits against a direct fp32 log-softmax reference, including a tail-precision regression. This resolves my blocker.

**Comment by chtruong814** [comment]
CI ran as part of a batch of PRs in the merge queue to consolidate CI runs
https://github.com/NVIDIA/Megatron-LM/pull/7462
https://github.com/NVIDIA/Megatron-LM/actions/runs/35306160723/job/105485956036

Tests passed, so will fast merge this.
